File: Editor/src/main/java/org/turbomedia/turboedit/editor/panes/file_browser/FileCard.java
package org.turbomedia.turboedit.editor.panes.file_browser;

import javafx.geometry.Insets;
import javafx.scene.CacheHint;
import javafx.scene.Cursor;
import javafx.scene.control.ContextMenu;
import javafx.scene.control.MenuItem;
import javafx.scene.image.Image;
import javafx.scene.image.ImageView;
import javafx.scene.input.MouseButton;
import javafx.scene.layout.Background;
import javafx.scene.layout.BackgroundFill;
import javafx.scene.layout.CornerRadii;
import javafx.scene.layout.VBox;
import javafx.scene.paint.Color;
import javafx.scene.text.Font;
import javafx.scene.text.FontSmoothingType;
import javafx.scene.text.Text;
import org.kordamp.ikonli.fluentui.FluentUiFilledAL;
import org.kordamp.ikonli.fluentui.FluentUiFilledMZ;
import org.kordamp.ikonli.javafx.FontIcon;
import org.turbomedia.turboedit.editor.events.EventSystem;
import org.turbomedia.turboedit.editor.events.EventType;
import org.turbomedia.turboedit.editor.events.ThemeChangedEventData;
import org.turbomedia.turboedit.editor.windows.file_browser.FileDetailsWindow;
import org.turbomedia.turboedit.shared.project.ProjectFile;

import java.io.IOException;

public class FileCard extends VBox {
    private final ContextMenu contextMenu = new ContextMenu();
    private final ProjectFile file;

    public FileCard(ProjectFile file) throws IOException {
        this.file = file;

        initContextMenu();

        var FONT_NAME = Font.getDefault().getName();

        EventSystem.RegisterListener(EventType.THEME_CHANGED, (dat) -> {
            var data = (ThemeChangedEventData) dat;

            Color color = switch (data.theme()) {
                case DARK -> Color.color(1, 1, 1, 0.05);
                case LIGHT -> Color.color(0, 0, 0, 0.05);
            };

            var background = new Background(new BackgroundFill(color, new CornerRadii(10), Insets.EMPTY));

            setBackground(background);
        });

        var resource = getClass().getResource("/test2.png");
        var stream = resource.openStream();
        var previewImage = new Image(stream);
        var preview = new ImageView(previewImage);

        preview.setPreserveRatio(true);
        preview.setFitWidth(160);
        preview.setCache(true);
        preview.setCacheHint(CacheHint.SPEED);

        var fileName = new Text(file.name());
        fileName.setFontSmoothingType(FontSmoothingType.LCD);
        fileName.setFont(Font.font(FONT_NAME, 12));

        setCursor(Cursor.HAND);
        getChildren().add(preview);
        getChildren().add(fileName);
        setSpacing(6);
        setPadding(new Insets(6));
        setOnContextMenuRequested((e) -> contextMenu.show(this, e.getScreenX(), e.getScreenY()));
        setOnMouseClicked((e) -> {
            if (e.getButton() == MouseButton.PRIMARY && e.isAltDown()) {
                new FileDetailsWindow(file);
            }
        });
    }

    private void initContextMenu() {
        var removeFile = new MenuItem("Remove File");
        removeFile.setGraphic(FontIcon.of(FluentUiFilledAL.DISMISS_24));

        var replaceFile = new MenuItem("Replace File");
        replaceFile.setGraphic(FontIcon.of(FluentUiFilledAL.LINK_24));

        var fileDetails = new MenuItem("File Details");
        fileDetails.setGraphic(FontIcon.of(FluentUiFilledAL.APPS_LIST_24));
        fileDetails.setOnAction((e) -> new FileDetailsWindow(file));

        var transcode = new MenuItem("Transcode");
        transcode.setDisable(true);
        transcode.setGraphic(FontIcon.of(FluentUiFilledMZ.VIDEO_CLIP_24));

        var transcribe = new MenuItem("Transcribe");
        transcribe.setDisable(true);
        transcribe.setGraphic(FontIcon.of(FluentUiFilledMZ.TEXTBOX_24));

        var proxySettings = new MenuItem("Manage Proxy Settings");
        proxySettings.setDisable(true);
        proxySettings.setGraphic(FontIcon.of(FluentUiFilledAL.DOCUMENT_EDIT_24));

        contextMenu.setAutoHide(true);
        contextMenu.setAutoFix(true);
        contextMenu.getItems().addAll(removeFile, replaceFile, fileDetails, transcode, transcribe, proxySettings);
    }
}

File: Editor/src/main/kotlin/turboedit/editor/panes/file_browser/FileCard.kt
package turboedit.editor.panes.file_browser

import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.geometry.Insets
import javafx.scene.CacheHint
import javafx.scene.Cursor
import javafx.scene.control.ContextMenu
import javafx.scene.control.MenuItem
import javafx.scene.image.Image
import javafx.scene.image.ImageView
import javafx.scene.input.ContextMenuEvent
import javafx.scene.input.MouseButton
import javafx.scene.input.MouseEvent
import javafx.scene.layout.Background
import javafx.scene.layout.BackgroundFill
import javafx.scene.layout.CornerRadii
import javafx.scene.layout.VBox
import javafx.scene.paint.Color
import javafx.scene.paint.Paint
import javafx.scene.text.Font
import javafx.scene.text.FontSmoothingType
import javafx.scene.text.Text
import org.kordamp.ikonli.fluentui.FluentUiFilledAL
import org.kordamp.ikonli.fluentui.FluentUiFilledMZ
import org.kordamp.ikonli.javafx.FontIcon
import turboedit.editor.events.EventSystem.EventAction
import turboedit.editor.events.EventSystem.registerListener
import turboedit.editor.events.EventType
import turboedit.editor.events.ThemeChangedEventData
import turboedit.editor.misc.StyleManager
import turboedit.editor.windows.file_browser.FileDetailsWindow
import turboedit.shared.project.ProjectFile

class FileCard(private val file: ProjectFile) : VBox() {
    private val contextMenu = ContextMenu()

    init {
        initContextMenu()

        val fontName = Font.getDefault().name

        registerListener(EventType.THEME_CHANGED, EventAction { dat: Any? ->
            val data = dat as ThemeChangedEventData?
            val color = when (data!!.theme) {
                StyleManager.Theme.DARK -> Color.color(1.0, 1.0, 1.0, 0.05)
                StyleManager.Theme.LIGHT -> Color.color(0.0, 0.0, 0.0, 0.05)
                else -> {}
            }

            val background = Background(BackgroundFill(color as Paint?, CornerRadii(10.0), Insets.EMPTY))
            setBackground(background)
        })

        val resource = javaClass.getResource("/test2.png")
        val stream = resource!!.openStream()
        val previewImage = Image(stream)
        val preview = ImageView(previewImage)

        preview.isPreserveRatio = true
        preview.fitWidth = 160.0
        preview.isCache = true
        preview.cacheHint = CacheHint.SPEED

        val fileName = Text(file.name)
        fileName.fontSmoothingType = FontSmoothingType.LCD
        fileName.font = Font.font(fontName, 12.0)

        cursor = Cursor.HAND
        children.add(preview)
        children.add(fileName)
        spacing = 6.0
        padding = Insets(6.0)
        onContextMenuRequested = EventHandler { e: ContextMenuEvent? ->
            contextMenu.show(
                this,
                e!!.screenX,
                e.screenY
            )
        }
        onMouseClicked = EventHandler { e: MouseEvent? ->
            if (e!!.button == MouseButton.PRIMARY && e.isAltDown) {
                FileDetailsWindow(file)
            }
        }
    }

    private fun initContextMenu() {
        val removeFile = MenuItem("Remove File")
        removeFile.graphic = FontIcon.of(FluentUiFilledAL.DISMISS_24)

        val replaceFile = MenuItem("Replace File")
        replaceFile.graphic = FontIcon.of(FluentUiFilledAL.LINK_24)

        val fileDetails = MenuItem("File Details")
        fileDetails.graphic = FontIcon.of(FluentUiFilledAL.APPS_LIST_24)
        fileDetails.onAction = EventHandler { e: ActionEvent? -> FileDetailsWindow(file) }

        val transcode = MenuItem("Transcode")
        transcode.isDisable = true
        transcode.graphic = FontIcon.of(FluentUiFilledMZ.VIDEO_CLIP_24)

        val transcribe = MenuItem("Transcribe")
        transcribe.isDisable = true
        transcribe.graphic = FontIcon.of(FluentUiFilledMZ.TEXTBOX_24)

        val proxySettings = MenuItem("Manage Proxy Settings")
        proxySettings.isDisable = true
        proxySettings.graphic = FontIcon.of(FluentUiFilledAL.DOCUMENT_EDIT_24)

        contextMenu.isAutoHide = true
        contextMenu.isAutoFix = true
        contextMenu.items.addAll(removeFile, replaceFile, fileDetails, transcode, transcribe, proxySettings)
    }
}
